### ISSUE TITLE
`:checkhealth`: Search `package.path` for health files.

### ISSUE THREAD

**Issue by mrcjkb** [neovim/neovim#26580]
### Problem

`:checkhealth` doesn't work for plugins that have been installed using luarocks and added to the `package.path` and the runtimepath.

The reason for this is that the [`copy_directories` field cannot include a "lua" directory](https://github.com/luarocks/luarocks/wiki/Rockspec-format), as it has a special meaning (installed to `lib`, instead of `share`)

### Expected behavior

A potential solution would be to search the `package.path` in the [`get_healthcheck_list`](https://github.com/mrcjkb/neovim-fork/blob/ef38fdfdc6c84abd8ce7be02eaf8edc91ebc7917/runtime/lua/vim/health.lua#L76) function:

```lua
local lua_lib_path = package.searchpath(p .. '.health', package.path)
if lua_lib_path then
  table.insert(paths, lua_lib_path)
end
```

I don't know if this would have any negative implications, or if there's a better way to solve this, which is why I'm opening this issue for discussion before I open a PR :smile:

**Comment by mrcjkb** [comment]
On the rocks.nvim side, we have implemented [a workaround](https://github.com/nvim-neorocks/rocks.nvim/pull/161/files).

Feel free to close this :smile: 

@justinmk I hope you don't mind me pinging you here, since you seem to be pretty active when it comes to luarocks + neovim.

**Comment by teto** [comment]
It looks a bit odd to look into luarocks packages for health.lua if they are not in rtp. In practice though it doesn't look hard to have sthg robust enough. I would say it might be premature to consider this for neovim, let's wait to see rocks.nvim adoption (I hope) and if the need becomes more widespread we can revisit ? especially now that you have a workaround (that doesn't look too annoying).

**Comment by justinmk** [comment]
> It looks a bit odd to look into luarocks packages for health.lua if they are not in rtp.

True. What is the reason for that? Why can't luarocks plugins be added to rtp?

**Comment by mrcjkb** [comment]
> Why can't luarocks plugins be added to rtp?

For a luarocks package, the `lua` and `lib` source directories have a special meaning, and are installed into a different directory than the directories specified with `build.copy_directories` in a rockspec.

It would be possible to hack the luarocks config to force a flat hierarchy (this is what nixpkgs does), but that is hacky and could potentially lead to issues for luarocks packages that aren't neovim plugins (e.g. ones that are written in C).

Besides that, for rocks.nvim, this would also mean that a plugin's dependencies' runtime (plugin, ftdetect, ...) scripts sourced during startup, which is something we don't want by design. We want a plugin's dependency's API (lua, autoload, ...) to be available, but its plugin scripts should only be sourced if it is explicitly installed by the user. 
That works quite well with the directory structure of luarocks packages. 

For now, we're happy with the symlinking workaround (which we now also use to make colorschemes and autoload scripts, etc. available), and haven't had any users report issues with it.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/41794#issuecomment-5601658338
